proposals/csharp-9.0/function-pointers.md

```diff
@@ -268,6 +268,75 @@ The better function member specification will be changed to include the followin
 
 This means that it is possible to overload on `void*` and a `delegate*` and still sensibly use the address-of operator.
 
+### Type Inference
+
+In unsafe code, the following changes are made to the type inference algorithms:
+
+#### Input types
+
+https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#input-types
+
+The following is added:
+
+> If `E` is an address-of method group and `T` is a function pointer type then all the parameter types of `T` are input types of `E` with type `T`.
+
+#### Output types
+
+https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#output-types
+
+The following is added:
+
+> If `E` is an address-of method group and `T` is a function pointer type then the return type of `T` is an output type of `E` with type `T`.
+
+#### Output type inferences
+
+https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#output-type-inferences
+
+The following bullet is added between bullets 2 and 3:
+
+> * If `E` is an address-of method group and `T` is a function pointer type with parameter types `T1..Tk` and return type `Tb`, and overload resolution
+of `E` with the types `T1..Tk` yields a single method with return type `U`, then a _lower-bound inference_ is made from `U` to `Tb`.
+
+#### Lower-bound inferences
+
+https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#lower-bound-inferences
+
+The following case is added to bullet 3:
+
+> * `V` is a function pointer type `delegate*<V2..Vk, V1>` and there is a unique type `delegate*<U2..Uk, U1>` such that `U` (or, if `U` is a type parameter,
+it's effective base class or any member of its effective interface set) is identical to, inherits from (directly or indirectly), implements (directly or
+indirectly), or has an implicit function pointer conversion to `delegate*<U2..Uk, U1>`.
```

Comment:
>or has an implicit function pointer conversion to `delegate*<U2..Uk, U1>`. [](start = 13, length = 74)

This looks suspicious. If `V` doesn't refer to an unfixed type parameter, there is nothing to infer. Otherwise, how can we check if a conversion exists? I think we should simply perform the inference with the goal to end up with `delegate*<U2..Uk, U1>` implicitly convertible to `delegate*<V2..Vk, V1>` after the inference, if that is at all possible. Therefore, we should probably state that calling conventions should match and ref modifiers in all positions should match. And then we simply perform inference through types in each position as specified below.

Reply:
I don't believe we need to state anything about ref and calling conventions: if `delegate*<U2..Uk, U1>` is implicitly convertible to `delegate*<V2..Vk, V1>`, then all the existing conversion rules apply, including those. I do thing you're right about the rest though.